### PR TITLE
Add support for Prometheus metrics

### PR DESCRIPTION
**Pull request by ruuda**
#### Problem

[Prometheus](https://prometheus.io/) is one of the most common tools used for monitoring software. At Chorus One we use it for all our monitoring, and almost all chains that we operate natively expose Prometheus metrics. Solana used to not be one of them, which made monitoring Solana a challenge.

One workaround is to rely on an external exporter, such as [certusone/solana_exporter](https://github.com/certusone/solana_exporter). We run this one, in combination with some custom scripts that follow the same approach of polling the RPC and exposing Prometheus metrics, and [our Lido maintenance daemon](https://github.com/ChorusOne/solido/blob/015d049fcbaffafe00518539ca76c7616720299a/cli/maintainer/src/daemon.rs) which also follows a similar approach, but is limited to monitoring very specific validators. However, an external exporter has some limitations and pitfalls:

* **Polling delay**. When the external watcher polls the RPC at time _t_, and Prometheus then polls the watcher at time _t_ + _d_, then Prometheus will record the sample at time _t_ + _d_, even though the value was actually produced at time _t_. This means that values are shifted on average by half of the external watcher’s polling interval, but the shift varies over time, and there is no way to know by how much. This is a problem if you want to compute whether one node is behind with respect to other nodes that you monitor. It is possible to fix this by exposing timestamps alongside the metrics. Our Lido exporter does this correctly, but `certusone/solana_exporter` doesn’t, and in addition you need to configure your Prometheus instance to respect these timestamps.
* **Stale metrics**. Due to the way most libraries that Prometheus work, when the exporter fails to poll the RPC, it continues to serve stale values, instead of stopping to expose the metrics. This makes the result in Prometheus misleading, because you might think that e.g. the validator was still online but not receiving new blocks, while in reality, the validator was down. In our Lido exporter we handle this case correctly, but in our ad-hoc scripts we had this bug. From a quick glance of `certusone/solana_exporter`, it is also susceptible to this bug, though I’m not an expert on Go so I might be wrong.
* **Inconsistent reads**. When an external exporter performs multiple RPC calls to gather its data, it might read from multiple slots, and end up with an inconsistent result. For example, if we want to monitor the amount of SOL in account A and B, an exporter might call `getAccountInfo(A)`, then a transfer from A to B happens, and then the exporter calls `getAccountInfo(B)`, and it exports the two values as `account_balance_sol{account="A"}` and `account_balance_sol{account="B"}`. Then when you query Prometheus for `sum(account_balance_sol)`, you will observe more SOL than exists. In our Lido exporter we handle this correctly by using `getMultipleAccounts`, but [it’s pretty involved and inefficient](https://github.com/ChorusOne/solido/blob/015d049fcbaffafe00518539ca76c7616720299a/cli/common/src/snapshot.rs), and doesn’t work for information that is not stored on-chain. (See also #17633.) An in-process exporter can just query the same bank multiple times and get a consistent result.
* **More moving parts**. There is an additional daemon that you need to run and keep up to date, and which can fail.
* **Not obvious which part fails.** If metrics are missing or flatlining, the problem could be in the validator, between the validator and the exporter, in the exporter, or between the exporter and Prometheus. Direct monitoring eliminates two of these cases.
* Not all values that we are interested in are available over RPC.
* Every instance of `solana-validator` needs to be paired with an instance of the exporter, as they will have different views of the current slot height etc.

To ease monitoring, we added native Prometheus metrics to the RPC server. We have been running this in production for about three months now, and it has greatly improved our visibility into how our nodes are running.

#### Summary of Changes

* Add a new http endpoint on the RPC interface, `/metrics`, which exposes metrics in Prometheus format.
* Add a new flag, `--monitor-vote-account`, to export details about a given vote account. We don’t export all vote accounts by default, as that is something user-controlled — we don’t want our Prometheus instance to slow down because it has to deal with 10,000 additional labels when a troll creates 10,000 vote accounts on the network. The validator’s own identity is always implicitly monitored.

Currently the following metrics are implemented:

```
# HELP solana_block_epoch Epoch of the current block
# HELP solana_block_epoch_slots_total The duration of the current epoch, in slots.
# HELP solana_block_epoch_start_slot The first slot in the current epoch
# HELP solana_block_slot Slot of the current block
# HELP solana_block_timestamp_seconds The block's UNIX timestamp, in seconds since epoch, UTC
# HELP solana_node_identity_balance_sol The balance of the node's identity account
# HELP solana_node_identity_public_key_info The node's current identity
# HELP solana_node_version_info The current Solana node's version
# HELP solana_snapshot_last_full_snapshot_slot The slot height of the most recent full snapshot
# HELP solana_snapshot_last_incremental_snapshot_slot The slot height of the most recent incremental snapshot
# HELP solana_validator_active_stake_sol The total amount of SOL actively staked to this validator
# HELP solana_validator_last_vote_slot The voted-on slot of the validator's last vote that got included in the chain
# HELP solana_validator_vote_account_balance_sol The balance of the vote account at the given address
# HELP solana_validator_vote_credits The total number of vote credits credited to this vote account
```

#### For others interested

We have a branch at https://github.com/ChorusOne/solana/tree/prometheus with these patches currently on top of 1.10.38. We keep this branch free of conflicts with the latest mainnet release and rebase it when necessary, but we don’t rebase it for every release. (We have some internal automation that applies the patch on top of the latest release when we build from source.)

For this pull request I rebased the patches on top of `master`, but I suppose if it gets accepted, it will still be a while before it makes it into a mainnet release, so we will continue to maintain our `prometheus` branch for a while.

#### Credits

Most of the implementation was done by @enriquefynn and @Szymongib. Parts of the code were adapted from our Lido maintainer daemon.

#### Next steps

This is an initial pull request with what works for us, but it needs some more work to be included in a release. In particular, we now expose metrics unconditionally at the RPC port, but nodes with public RPCs might not want that, so we should gate this behind a default-off command-line flag. Before we go there though, I’d like to gauge if you are interested in accepting this upstream at all. The changes are pretty isolated and not invasive to the rest of the codebase, and there is no performance impact at runtime as long as you don’t query the endpoint. And even then, the impact is no worse than that of a regular RPC call.

We are also planning to add more metrics, also from sources that are not available over the RPC. Solana already records these in many places, but exposing them will be a bit more invasive than the current metrics.

### CODE REVIEW

**Comment by leoluk** [comment]
Yes please! I'd love to get rid of solana_exporter

**Comment by linuskendall** [comment]
This would be great, I've already made a Triton build with this and will be testing it on some of our nodes. We have some RPC metrics that we might look to contribute after this is merged as a separate PR to these crates.

**Comment by ruuda** [comment]
We have a 1.13-compatible version of this at https://github.com/ChorusOne/solana/tree/prometheus-1.13.

**Comment by nyetwurk** [comment]
PROM is hard requirement for some of our observability roadmap; we're investigating doing this internally, but this would help immensely. Please consider merging asap

**Comment by linuskendall** [comment]
We've been running this internally and looks good. We have a few patches we'll try to upstream into this patch too (particularly if it gets merged). Would be so much easier if this was in main though!

**Comment by pires** [comment]
This is definitely something we're looking forward to have merged, as well.

**Comment by ruuda** [comment]
For those interested in this pull request, we continue to maintain up to date patches at https://github.com/ChorusOne/solana/tree/prometheus-1.13. We’ll soon update it for 1.14, and we plan to maintain these patches as long as they are not integrated upstream. Our fork is open to pull requests, in case anybody has additional metrics to add.

File: prometheus/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+[package]
```
Comment (dovahcrow) [inline prometheus/Cargo.toml:1]: Maybe I'm lazy, but why not use the Prometheus rust [crate](https://crates.io/crates/prometheus)?

**Comment by ruuda** [comment]
We have a 1.14-compatible branch of this at https://github.com/ChorusOne/solana/tree/prometheus-1.14.

**Comment by willhickey** [comment]
This repository is no longer in use. Please re-open this pull request in the agave repo: https://github.com/anza-xyz/agave

**Comment by nyetwurk** [comment]
@ruuda did you already open one? We'd like this merged as well.

**Comment by ruuda** [comment]
I’m waiting to see how that fork turns out, the process of migrating the repository looks like a big mess, and even if most of the developers moved, I haven’t seen any announcement from Solana Labs, which worries me. I haven’t yet thought very deeply about how trustworthy the fork is.

Given that there has been no response from the developers on this for 1.5 years despite clear interest from node operators and several attempts from our side to reach out, I don’t have high hopes that it would get merged in the new fork.

We still maintain these patches, the most recent branch is https://github.com/ChorusOne/solana/tree/prometheus-1.17.17, and those commits rebase without conflict on top of the latest 1.17 releases. We have internal automation that automatically creates a tag with the patches applied for every upstream tag in _this_ repository, we haven’t yet evaluated whether we want to switch to the fork.

**Pull request closed without merging**